### PR TITLE
Select dihedrals faster

### PR DESCRIPTION
**Pull request by lilyminium**
Fixes (part of) #2671

Changes made in this Pull Request:
 - made phi/psi_selection longer, but faster with manual boolean array matching of atom names
 - made Ramachandran really long and ugly, but really faster with manual boolean array matching to atom names
 - no longer hardcodes phi/psi selection names in Ramachandran

The original ``Ramachandran.__init__`` with the GRO/XTC files was [180 s on my laptop.](https://github.com/MDAnalysis/mdanalysis/issues/2671#issuecomment-636708041) When I changed only the `phi_selection` code (not `psi_selection`), it dropped to 122 s. When I changed the Ramachandran code, it dropped to 0.25 s. Most of that is selecting the protein.

Screenshot 2020-06-01 at 9 23 10 PM

~How much testing does this need?~ lots

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

To do
-----
- [x] update omega_selection and chi1_selection too
- [x] add `c_name`, `n_name` etc. as kwargs to those, it's annoying that the names are hard-coded in (edit: will *remove* pending *dis*approval from another dev)
- [x] add the group selection methods to `ResidueGroup`
- [x] update `Ramachandran.__init__` to use the group selection methods

Edit: Ramachandran is no longer in the slowest 50 tests.
```
========================== slowest 50 test durations ===========================

39.76s call     MDAnalysisTests/analysis/test_density.py::Test_density_from_Universe::test_notwithin

31.18s call     MDAnalysisTests/analysis/test_density.py::TestNotWithin::test_within

30.83s call     MDAnalysisTests/analysis/test_density.py::TestNotWithin::test_not_within

29.70s call     MDAnalysisTests/analysis/test_density.py::TestNotWithin::test_has_DeprecationWarning

29.58s call     MDAnalysisTests/analysis/test_hbonds.py::TestHydrogenBondAnalysis::test_true_traj

29.21s call     MDAnalysisTests/analysis/test_hbonds.py::TestHydrogenBondAnalysisPBC::test_true_traj

29.15s call     MDAnalysisTests/analysis/test_hbonds.py::TestHydrogenBondAnalysisHeuristic::test_true_traj

29.11s call     MDAnalysisTests/analysis/test_hbonds.py::TestHydrogenBondAnalysisHeavy::test_true_traj

29.11s call     MDAnalysisTests/analysis/test_hbonds.py::TestHydrogenBondAnalysisHeavyFail::test_true_traj

21.78s call     MDAnalysisTests/visualization/test_streamlines.py::test_streamplot_2D

14.44s call     MDAnalysisTests/analysis/test_hole2.py::TestHoleAnalysis::test_context_manager

12.24s call     MDAnalysisTests/coordinates/test_pdb.py::test_conect_bonds_all

10.71s call     MDAnalysisTests/analysis/test_density.py::Test_density_from_Universe::test_update_selection

9.67s call     MDAnalysisTests/analysis/test_density.py::TestDensityAnalysis::test_run[dynamic]

8.56s call     MDAnalysisTests/coordinates/test_gro.py::TestGROLargeWriter::test_writer_large

8.44s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis

8.34s call     MDAnalysisTests/analysis/test_gnm.py::test_closeContactGNMAnalysis_weights_None

7.49s call     MDAnalysisTests/analysis/test_encore.py::TestEncore::test_ces_error_estimation_ensemble_bootstrap
```

### CODE REVIEW

**Comment by richardjgowers** [comment]
This looks cool. Would it make sense to put the new logic as a ResidueGroup
method? (Psi_selections)

**Comment by lilyminium** [comment]
@richardjgowers that's a great idea, thanks! While we're looking at `phi/psi_selection`, though, and given the recent discussion on residx vs resids, *should* we be using `segid` and `resid` to select atoms? (I kept this logic in the manual matching for now). I can see the importance of `resid` when the protein is missing a chunk in the middle, but how should `segid` be treated? If multiple segments are called the same thing, are they supposed to be treated as the same segment?

**Comment by richardjgowers** [comment]
I’m not enough of a bio person to really answer. I do know that resindices
make no promises on ordering, it’s just the unique number we give to each
residue.

**Comment by lilyminium** [comment]
`Ramachandran.__init__` is now ~25 lines shorter but takes 0.32 s on my laptop :-( I'll let other people decide if it's worth the slowdown.

File: package/MDAnalysis/analysis/dihedrals.py [hunk 1]
```diff
@@ -265,12 +265,21 @@ class Ramachandran(AnalysisBase):
...
     def __init__(self, atomgroup, c_name='C', n_name='N', ca_name='CA',
-                 **kwargs):
+                 check_protein=True, **kwargs):
         """Parameters
         ----------
         atomgroup : AtomGroup or ResidueGroup
             atoms for residues for which :math:`\phi` and :math:`\psi` are
             calculated
+        c_name: str (optional)
+            name for the backbone C atom
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/dihedrals.py:275]: specify defaults for these in the docstring

Comment (richardjgowers) [reply package/MDAnalysis/analysis/dihedrals.py:275]: and maybe add a line showing how they can be used

File: package/MDAnalysis/analysis/dihedrals.py [hunk 2]
```diff
@@ -265,12 +265,21 @@ class Ramachandran(AnalysisBase):
...
+        n_name: str (optional)
+            name for the backbone N atom
+        ca_name: str (optional)
+            name for the alpha-carbon atom
+        check_protein: bool (optional)
+            whether to raise an error if the provided atomgroup is not a 
+            subset of protein atoms
 
         Raises
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/dihedrals.py:284]: needs a `versionchanged::` for the extra kwargs

File: package/MDAnalysis/core/topologyattrs.py [hunk 3]
```diff
@@ -655,39 +655,72 @@ def _get_next_residues_by_resid(residues):
...
-            The next residue in Universe. If not found, the corresponding 
-            item in the list is ``None``.
+            List of the next residues in the Universe, by resid and segid. 
+            If not found, the corresponding item in the list is ``None``.
 
         .. versionadded:: 1.0.0
         """
         u = residues[0].universe
-        nxres = rview = np.array([None]*len(residues), dtype=object)
+        nxres = np.array([None]*len(residues), dtype=object)
```
Comment (richardjgowers) [inline package/MDAnalysis/core/topologyattrs.py:664]: fwiw, seems like a candidate for atomgroupgroup #1861

File: package/MDAnalysis/analysis/dihedrals.py (hunk 1 above)
Comment (lilyminium) [reply package/MDAnalysis/analysis/dihedrals.py:275]: I remember discussing this with @orbeckst and coming to the agreement that defaults are in the function signature anyway, so it's easier for docs not to diverge if they're not specified in the docstring. Happy to go either way though

**Pull request merged**